src/main/java/com/mongodb/kafka/connect/source/MongoSourceTask.java

```diff
@@ -309,18 +330,25 @@ private void setCachedResultAndResumeToken() {
             cursor = createCursor(sourceConfig, mongoClient);
         }
 
-        try {
-            BsonDocument next = cursor.tryNext();
-            if (next == null && cursor.getServerCursor() == null) {
-                cursor.close();
-                cursor = null;
+        if (cursor != null) {
+            try {
+                BsonDocument next = cursor.tryNext();
+                // The cursor has been closed by the server
+                if (next == null && cursor.getServerCursor() == null) {
+                    invalidatedCursor = true;
+                    cursor.close();
+                    cursor = createCursor(sourceConfig, mongoClient);
+                }
+                return Optional.ofNullable(next);
+            } catch (Exception e) {
+                if (cursor != null) {
+                    cursor.close();
+                    cursor = createCursor(sourceConfig, mongoClient);
```

Comment:
There was some other cursor error - close and create a new cursor.

Reply:
Reverted so if creating a cursor errors - it will be caught correctly in the next loop.